util/clusterauth/clusterauth.go

```diff
@@ -215,39 +218,121 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
 	return GetServiceAccountBearerToken(clientset, ns, ArgoCDManagerServiceAccount)
 }
 
-// GetServiceAccountBearerToken will attempt to get the provided service account until it
-// exists, iterate the secrets associated with it looking for one of type
-// kubernetes.io/service-account-token, and return it's token if found.
+// GetServiceAccountBearerToken determines if a ServiceAccount has a
+// bearer token secret to use or if a secret should be created. It then
+// waits for the secret to have a bearer token if a secret needs to
+// be created and returns the token in encoded base64.
 func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
-	var serviceAccount *corev1.ServiceAccount
+	secretName, err := getOrCreateServiceAccountTokenSecret(clientset, sa, ns)
+	if err != nil {
+		return "", err
+	}
+
 	var secret *corev1.Secret
-	var err error
-	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+	err = wait.PollImmediate(500*time.Millisecond, 30*time.Second, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), ClientGoRequestTimeout)
+		defer cancel()
+		secret, err = clientset.CoreV1().Secrets(ns).Get(ctx, secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to get secret %q for serviceaccount %q: %w", secretName, sa, err)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get token for serviceaccount %q: %w", sa, err)
+	}
+
+	_, ok := secret.Data["token"]
+	if !ok {
+		return "", fmt.Errorf("secret %q for serviceaccount %q did not have a token", secret.Name, sa)
+	}
```

Comment:
Do we want this inside the poll loop? As it's currently written, the poll will only ever run once, because there is no "still waiting" condition. I think that this check can go inside, and we can just return `false, nil` to indicate that the poll logic should retry.

Reply:
Originally doing this was causing panics when I ran the test locally. This is due to my vscode setup where running tests were being timed out after 30s. The downside of moving this into the poll loop is the clusterauth tests now take 60s to run: `ok  	github.com/argoproj/argo-cd/v2/util/clusterauth	65.522s	coverage: 74.7% of statements`.

This is from tests that were looking at how this would behave in a scenario where the token key never populates. It's working in CI, but might be not so great running locally.

Reply:
Could we parameterize the timeout so it can be overridden for tests?

Reply:
Yeah, I can add to InstallClusterManagerRBAC/GetServiceAccountBearerToken.

Reply:
Added a 5 second timeout for tests.